Initial implementation and tests

RangerMauve opened the pull request:
Fixes # .

Changes proposed in this pull request:
-
-
-

@consento-org/maintainer

RangerMauve committed:
Initial progress

RangerMauve committed:
Initial implementation and tests

martinheidegger committed:
removing .map files

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+
+  processFeeds(): boolean {
+    let hasProcessed = false;
+    let hasResponded = false;
```
`hasResponded` is set but not used.

RangerMauve replied:
Is there a way to get typescript to detect stuff like that? I feel like I might have to add eslint into the mix.

Having three tools to do linting feels a bit daunting. 😂 Parsing all the code at least 3 times per lint.

RangerMauve replied:
The variable was supposed to be used later. Good catch!

dkastl replied:
What do you need Typescript for to detect this if you have Martin? ;-)

martinheidegger replied:
Note: my IDE uses code colors to highlight unused variables.

RangerMauve replied:
Yeah, I usually have eslint cover unused variables and set up my tests to make sure. 😁

I guess eslint will be necessary after all. In which case I think I'd rather remove prettier from the workflow.

RangerMauve replied:
It'd be nice if the typescript type checking was inside eslint, too. I'll see if that's an option. 😁

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+          const req = this.getRequest(item.id);
+          req.addResponse(id, item);
+          let neededSignatures = this.knownMembers.length;
+          if (req.operation === "remove") neededSignatures--;
```
Shouldn't this be always `n-1`?

RangerMauve replied:
If we do `n-1` can we avoid ambiguity? Like, if we have A,B,C,D,E, if A,B,C have signed, and D/E aren't connected to each other / accept, they'll both think they're the final needed signature and cause a fork.

martinheidegger replied:
If `n-1` doesn't work in that case it doesn't work in the other case, as two devices can mutually remove each other, it would also create that ambiguity.

martinheidegger replied:
We had a conversation about allowing forks but maybe resolving equal forks?! In example: D & E approve R1 creating Forks R1D and R1E and if they sync and notice equality they decide for the alphabetically lower one (R1D).

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+          if (req.operation === "remove") neededSignatures--;
+          const maxDenied = this.knownMembers.length - neededSignatures;
+          const state = req.getState(neededSignatures, maxDenied);
+          if (req.getState(neededSignatures) === "ready") {
```
This should probably be `state === "ready"` ?

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+  return (item as MergeRequest).toMerge !== undefined;
+}
+
+function isWho(item: Request): item is WhoRequest {
```
When I first read `isWho` I was pretty confused, same for `WhoRequest` Maybe rename?

RangerMauve replied:
Renamed

martinheidegger commented on test/index.ts:
```diff
@@ -0,0 +1,115 @@
+
+  t.end();
+});
+
```
There are a few essential tests missing here:

1. one member does an operation while another removes that member: who wins?
2. Multiple requests processed at once where the first adds a member (preventing the second to go through because more request are needed)
3. two members start an add operation in parallel, what happens?

RangerMauve replied:
Adding TODOs to the README to keep track of these, will work on them this week

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+
+    if (hasProcessed) {
+      const otherRun = this.processFeeds()
+      return hasProcessed || otherRun;
```
Ugh, endless recursion possible, right? 😅

RangerMauve replied:
Err, what's the condition that causes it?

martinheidegger replied:
Nothing at the moment, but my eyes look with stress 💦 on the code and I would be constantly worried if it introduces endless recursions.

RangerMauve replied:
Is there a different approach I should take here? The current structure recurses until all the feeds have been processed (more specifically, until no items are processable).

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+    // Get new data from members to knownFeeds
+    for (const id of Object.keys(member.knownFeeds)) {
+      const feed = member.knownFeeds[id];
+      const ownCopy = this.getFeedFor(id);
```
There should be a validate template here to validate if the feed data that is added keeps being added. I think this is important to figure out even in an early stage because it allows us to figure out if we **can** identify a valid from a corrupt feed.

RangerMauve replied:
Would this require the signvector stuff?

martinheidegger replied:
The sign-vectors are part of this too. Yes, but there is a much simpler case: if there is a feed of a member that has been removed or was never added. Also, there is a thin possibility that the sign-vectors fail (and a fork is created). i.e. malfunctioning of a client: Fork detection should also be part of it.

Note: The central log may be a tool to make it easier/possible to detect member-feed forks.

RangerMauve replied:
Hmm, mind pseudocodeing how the verification would look and what the test for forking would look like?

martinheidegger replied:
```javascript
function isIncomingFeedForked (localFeed, feed) {
  // Look at the merkle tree and see if the feed was forked at a previous step
}

function isIncomingFeedSignVectorCorrect (feed, startIndex) {
  let vector = createSignVectorFor(feed.get(startIndex))
  for (let i = startIndex+1; i<feed.length; i++) {
    if (!signVector.validate(feed.get(i))) return false
  }
  return true
}

function isIncomingFeedCorrect (feed) {
  // testing "correct-ness" of requests, i.e. proper msgpack/protobuf object; size limit not exceeded; removes feed that actually existed. Does the signature exist for the request? etc.
}

function verifyFeed (localFeed, feed, startIndex, peer) {
  if (isIncomingFeedForked(localFeed, feed)) {
    // Red alert
    requestRemoveMember({ cause: 'fork-detected' })
    return false
  }
  if (!isIncomingFeedSignVectorCorrect(feed, startIndex)) {
    addToBlackList(peer)
    return false
  }
  if(!isIncomingFeedCorrect(feed)) {
    // orange alert
    requestRemoveMember({ cause: 'incorrect-client' })
    return false
  }
  return true
}
```

RangerMauve replied:
Sweet, thank you!

martinheidegger commented on test/index.ts:
```diff
@@ -0,0 +1,115 @@
+
+tape("Able to add a member by ID and sync", (t) => {
+  const member = new Member();
+  const other = new Member({ initiator: member.id });
```
This adds the requirement that the other member knows the initiator at start. QR code for `member` has to be given to `other` and for `requestAdd` the `other`'s QR-code has to be given to `member`. It should be possible to reduce this so that only the `requestAdd(other.id)` is necessary ?!

RangerMauve replied:
My assumption was that only known members were allowed to request an add, therefore if a new peer was to be added, they'd need to know the ID of the first member in order to traverse the history and get to the latest state.

Ties into what we talked about regarding using just member feeds for representing history and traversing them instead of a group feed.

martinheidegger replied:
> need to know the ID of the first member

The ID of the first member should be in the first _(and only?)_ block of the group feed, no?

But beside that: The workflow needs indeed to be `group-id` → new member → `new-member-id` → existing member.

RangerMauve replied:
Hmm, yeah, my current code doesn't account for the group feed or how the initial member is found. Was hoping to punt that for later and focus on the request flow.

martinheidegger replied:
It is okay to make this a memo.

martinheidegger commented on test/index.ts:
```diff
@@ -0,0 +1,115 @@
+tape("Able to initialize a member", (t) => {
+  const member = new Member();
+
+  member.processFeeds();
```
Shouldn't processFeeds() be part of the syncing process? What advantage do we get of separating those?

RangerMauve replied:
Main thing here was the auto-accept function meant that processing feeds would trigger member adds, and I was hoping to have more control of when that happened. I can refactor to work around that so that control of accepting could be outside of syncing feeds.

Other consideration was that I wanted a place for members to process requests and responses that they generated themselves 1. for the initial member case and 2. to remove the need for extra logic for that bit. Down to add proccesFeeds at the end of sync, though.

martinheidegger replied:
> I can refactor to work around that so that control of accepting could be outside of syncing feeds.

Structurally it is probably for the best, though I don't think we need to use manual confirmation in our first real scenario. The reason why I brought this up is because in the sync process it may be possible to avoid some chance of forks by having one member sync first and process the entries before syncing everything - including the need entries - with the other peer.

RangerMauve replied:
Yeah, I'm thinking that I'll have a `getPendingRequests()` method and a `acceptPendingRequests()`

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+  }
+
+  finish() {
+    this.finished = true;
```
Memo: This needs to be established in the final solution using a log statement.

RangerMauve replied:
Could you elaborate on this point?

martinheidegger replied:
Setting a `finished = true` statement works for this simulation but in an actual case we should probably have a log statement that looks like `{ request: 'abcd', finished: true }` in order to recover the state from the log itself.

RangerMauve replied:
Hmm, not 100% sure what you mean regarding recovering state from the log. If we don't persist request state anywhere we'd need to traverse the entire log again and that means that `finished` would be derived from there again.

martinheidegger commented on src/member.ts:
```diff
@@ -0,0 +1,358 @@
+  }
+
+  isCancelled(): boolean {
+    return this.signatures[this.from]?.response === ("cancel" as ResponseType);
```
This should work without the casting.

RangerMauve replied:
Thank you. Still getting used to what is and isn't limited by TS. 😅

RangerMauve committed:
Some revisions based on review

RangerMauve commented:
Working on refactoring to change how processFeeds works, and then I'll add more tests

martinheidegger commented:
Without having a central feed its quite challenging for me to figure out _if_ a fork happened. Is there a good way to verify that without a central log?

Another thing: When adding feeds, the information should not just contain their ID's but also the verification vector for the first block

RangerMauve commented:
Without a central feed there _is_ no forking.

Since everyone needs to agree on a history and you derive members by traversing each others feeds and verifying that signatures happened. If someone tires to fork their own member feed (assuming they bypass the signvector layer of security) then at most they can prevent new peers from getting a valid history (until the fork is detected and the invalid feed is purged).

Agreed on the verification vector stuff. I'll look at incorporating sign vectors after this initial refactor.

RangerMauve commented:
TODO:

- Add sign vectors
- Verify sign vectors in sync
- When removing specify index to remove at so that they won't be processed after
- Add tests for forks
- Add tests fo rconflicting requests
- Add tests so that external peers always process feeds in the same way

RangerMauve committed:
Separate out accepts, add test for concurrent requests

RangerMauve committed:
Add happy path for adding members and syncing with an external peer

RangerMauve commented:
Added some sims, so far so good.

Talking with Martin last week, we determined we could be running into issues if we allow members to have multiple in-flight requests.

Gonna write up some thoughts on how that works and make some diagrams of what it would look like.

martinheidegger committed:
moved to eslint & standard

martinheidegger committed:
adding husky precommit lint fix

martinheidegger committed:
adding husky precommit lint fix

RangerMauve committed:
Add dist to ignore, fix up test

RangerMauve committed:
Merge branch 'initial' of github.com:consento-org/permissions-system-sim into initial

RangerMauve committed:
Make linter happy

martinheidegger committed:
refactoring tests

martinheidegger committed:
refactor: returning the accepted requests

martinheidegger committed:
fix: two members trying to remove each other

martinheidegger commented:
When adding https://github.com/consento-org/permissions-system-sim/pull/1/commits/4651ac0925e24363bee028a12b5dc6a001b558e7 it came to me that the one and two members situation not only is special, but also that the we have at least three options on how to deal with these.

1. Don't allow two members: The first member **has** to add two members at once. In practice this would mean having a virtual member that is basically a QR code.
2. Treat one member (either first or second) with higher power, making it possible for the stronger to remove the weaker but not the other way around.
3. (current) Require both members to sign the removal of one. Meaning that you can't "loose" either of the initial two members.

martinheidegger committed:
fix: proper naming for the members

martinheidegger committed:
rename pending-requests to unsigned-requests and limit to 1 req per member.

Also renames acceptPending to signUnsigned. getPendingRequests() still exists but that show all requests that have not been completely signed. It also adds an intermediary method: getActiveRequests which contains only the first pending request for each known member.

- getPendingRequests ... all requests that have not been signed
- getActiveRequests ... one pending requests for each member
- getUnsignedRequests ... active requests that have not yet been signed by me.

martinheidegger commented:
To figure out the complexity of the system I would like to propose a "possible action" api:

```javascript
const a = new Member()
const b = new Member(a.groupId)
a.possibleActions // [sealGroup, addMember]
b.possibleActions // [] - not a member of the group
a.possibleActions.addMember(b.id)

a.possibleActions // [syncB, requestAdd, revertAdd - as its not synced with B yet]
a.possibleActions.syncB()

a.possibleActions // [removeB, addMember, syncB]
// etc.
```

By having a list of "possible actions" we can create an operational matrix. Create branch test after each operation where we see if all possible options on each member are exhausted by the tests.

martinheidegger committed:
test: adding test for what happens if both members remove each other

RangerMauve commented:
> When adding [4651ac0](https://github.com/consento-org/permissions-system-sim/commit/4651ac0925e24363bee028a12b5dc6a001b558e7) it came to me that the one and two members situation not only is special, but also that the we have at least three options on how to deal with these.
> 
> 1. Don't allow two members: The first member **has** to add two members at once. In practice this would mean having a virtual member that is basically a QR code.
> 2. Treat one member (either first or second) with higher power, making it possible for the stronger to remove the weaker but not the other way around.
> 3. (current) Require both members to sign the removal of one. Meaning that you can't "loose" either of the initial two members.

Is the current really requiring both members to sign? I thought that removes were `n - 1`.

Personally I don't think adding two members at once is going to work well for UX.

I really like option 2 and think that "order that they were added in" would be a good way to rank power. I think that rather than stronger or weaker, it'd be like "who's been here longer". Though I can totally see the case where the older member will allow themselves to be removed through inaction. 🤷 But I guess that's an interesting case for forked feeds.

On another note,, I'm not sure I understand the possibleActions yet. What sorts of values can they have and when will those values change? The thing throwing me off is `syncB` Does that mean "sync request B"?

RangerMauve commented:
BTW: Flow charts: https://hackmd.io/xiiC_CmxS1aV_aHO5qR4pw?both

Side note, in the comments on the flowcharts @martinheidegger  mentioned "How does Bob know Alice".

I was imagining that when the group is being created, the initial member would somehow get encoded into that groups identifier. That way you'd be able to start traversing the history of the initial member and find all the subsequent ones through adds.

Would that make sense? What if the group ID was the initial member ID even?

Side note, what should we do when we detect forks when there's only one peer online that can vouch that there's a different history available from what the forker provided? I think this is mostly a vulnerability with the initial member. 🤔

martinheidegger commented:
> Is the current really requiring both members to sign? I thought that removes were n - 1.

I added this exception in https://github.com/consento-org/permissions-system-sim/pull/1/commits/4651ac0925e24363bee028a12b5dc6a001b558e7. If there are only two members: both are required. 

> Personally I don't think adding two members at once is going to work well for UX.

Adding two members would look like: you try to add a second member and at the same time you need to create a "virtual member" which is an additional password/secret key/qr-code that needs to be virtually restored in order to remove one of the two members.

> On another note,, I'm not sure I understand the possibleActions yet. What sorts of values can they have and when will those values change? The thing throwing me off is syncB Does that mean "sync request B"?

With `possibleActions` we could track if our tests actually tested all possible combinations. Action `A` causes `x` possible Actions per `y` other members and in the end we can track if we tested out all paths in the test somehow.

They could also be part of a "verification" tool: we only accept requests and signatures if they are actually "possible".

RangerMauve commented:
Who would then control the virtual member? How would that be stored in any way that's different from either one member having priority on removes or needing both members to consent to a removal?

Is possibleActions kinda like the state checking code that we have now but a bit more declarative?

My gut there feeling is that defining it through actions will quickly grow the number of things that need to be checked vs something like `b.isMember` and `b.seesRequest('b')`. I don't feel super strongly about it though. 😁

martinheidegger commented:
Expanding on my point:

The test framework could do autobranching:

```javascript
const executedOps = new Set()

function testPossibility (members, maxMembers, operation) {
  members = cloneMembers(members) // Cloned state
  validResult(operation(members, maxMembers)) // lets not exceed a max complexity here
  validateState(members) // Are the members (still) in a valid state, no forks?
}

function testPossibilities(members, maxMembers) {
  const possibleOps = combinePossibilities(members)
  for (const op of possibleOps) {
     if (executedOps.has(op.id)) {
       continue
     }
     executedOps.add(op.id)
     testPossibility(members, maxMembers, op)
  }
}

testPossibilities([new Member()], 7)
```

RangerMauve commented:
Ah interesting. So the possibilities would try to generate every possible case and detect conflicting states or locks?

martinheidegger commented:
> Who would then control the virtual member? How would that be stored in any way that's different from either one member having priority on removes or needing both members to consent to a removal?

The UX can still look similar as in the case 2.) the difference would just be that it gives the control and responsibility to the implementer. The implementer could make this in a way that this virtual member (which could be created in any case!) is simply stored at the same place as the first member or the second. Making the 1st member or the 2nd member stronger, depending on where that information is stored.

martinheidegger commented:
> Ah interesting. So the possibilities would try to generate every possible case and detect conflicting states or locks?

Yes.

RangerMauve commented:
@martinheidegger Regarding the virtual member, is this so that we can allow flexability to leave it to the app to choose which member has precedence?

Could the same thing be done with a flag in the add request for the second member? A member is a lot of state to keep track of and I feel like it'd make it a lot harder to implement in the client side (does a virtual member have a new replication swarm and ID?)

Also forcing there to be two members at once seems to add a bit of complexity. Are they a new type of request that has a `who` array instead of a `who` field? Is it two requests? How would that play with the one request per peer stuff.

What I'm trying to get at is that if we really need that flexibility there might be simpler ways to get it. 😅

RangerMauve commented:
Regarding `possibleActions` I think that's a good idea, though it might be good to have some hand coded cases too so that we can get a better grasp on the details of what's happening. Testing all possible outcomes feels like a stress testing thing to have once the code is more stable rather than early experimental parts which can change drastically.

martinheidegger commented:
> Could the same thing be done with a flag in the add request for the second member? A member is a lot of state to keep track of and I feel like it'd make it a lot harder to implement in the client side (does a virtual member have a new replication swarm and ID?)

Could be done with a flag, the point about the flag is that we need to add a <if> statement while the the other approach would push that responsibility to the user. _Note: the virtual member behaves and acts just like a regular member in our system._

> Also forcing there to be two members at once seems to add a bit of complexity. 

The other option would be that the creation of a group already requires 3 members to start off and can never have less than three. (First block contains 3 members instead of 2). Removing one member of a 3 member group could be equal to "close group".

RangerMauve commented:
I'm still not sure what the benefit of pushing complexity onto users is here. 😅 It feels like it'd lead to the same dynamics as having two users even if there's an extra virtual user.

Needing a minimum of 3 members also feels like a rough spot for UX since I could imagine wanting to be able to use the system on a single device / add my other devices as I set them up. Or are we saying consento shouldn't be used there and should only be used for cases where you already have a group?

martinheidegger commented:
Yeah, i cant really defend that option anymore.

RangerMauve commented:
Issue I'm trying to think through right now is members resolving to different orders for knownMembers due to concurrent requests, and how that might be different from what external members would see. 😵

This'd be a problem if we use "who was first" for precedence in removes.

RangerMauve commented:
Here's a diagram describing the problem. https://hackmd.io/xiiC_CmxS1aV_aHO5qR4pw?both#Concurrent-Requests---Problem

Ideally it'd be good to find a way to get members that resolve to something other than the "canonical" ordering to recover. Maybe they need to re-calculate after seeing a new accepted request?

RangerMauve committed:
Add test for conflicting requests

RangerMauve commented:
Just added a test that demonstrates the issue:

A is waiting for E to sign Request D
C is waiting for D to sign Request E
E and D think everything is fine and converge to the same set where they're both visible.

Is this where we should star thinking about request merges?

martinheidegger commented:
One way how I could see this being resolved: Order members by ID in case of ambiguity (It is possible to identify the ambiguity)

RangerMauve commented:
HMMMM. I think I'm going to sleep on it for now.

Some thoughts:

- Detecting the ambiguity we detect concurrent requests would be good. (test shows they all see 2 pending requests at some point)
- Would it make sense to try to enforce the order that E and D come to?
- When making a request, could we use a vector clock and the set of members that need to sign it to ensure stuff like this?

martinheidegger commented:
> I think I'm going to sleep on it for now.

Sleep well!

> Would it make sense to try to enforce the order that E and D come to?

Well: resolving the ambiguity is in a sense "enforcement", isn't it?!

> When making a request, could we use a vector clock and the set of members that need to sign it to ensure stuff like this?

More information will make it easier for us to sort (i.e. we could use the accept order of all members) and it will clear up many ambiguities, but I am not sure if we can remove all of them.

RangerMauve commented:
- Integrate Martin's HLC code
- Add timestamps to each Request and Response
- Sync timestamps with peers?
- When seeing a request, check what the members are at that time and add to the request time

RangerMauve committed:
Progress on HLC, track knownMembers at request time

RangerMauve commented:
Added the clock, and did some refactoring.

Here's the main behavior that's changed:

- Requests and Responses are tagged with HLC clocks
- Responses get added to a list when they've been `finished`
- `knownMembers` is dynamically calculated from the finished responses (sorted by the time they were finished using HLC clocks)
- Requests are tagged with the knownMembers at the time they were seen (will change)
- If there are two concurrent requests, they will both be satisfied when the when the knownMembers at the time sign off.

Some more stuff to do:

- Validate there aren't any backwards time jumps in people's feeds
- Instead of passing knownMembers at the time, pass a getter so it can be generated as new requests get synced.

martinheidegger committed:
chore: dependency update & lint fix

martinheidegger committed:
chore: adding coverage information

martinheidegger committed:
Adding States util

martinheidegger committed:
first permissions

martinheidegger committed:
making sure that the first member can only add itself

martinheidegger committed:
preventing requests from unknown members

martinheidegger committed:
first request needs to be add request

martinheidegger committed:
first operation can not be a response

martinheidegger committed:
first member can add second member directly

martinheidegger committed:
with more than two members we can not simply add members

martinheidegger committed:
added request states

martinheidegger committed:
adding missing member amount check

martinheidegger committed:
refactor: simplifying logic

martinheidegger committed:
memo timestamps

martinheidegger committed:
other members may have older timestamps

martinheidegger committed:
permissions keeps internal clock

martinheidegger committed:
response for unknown request

martinheidegger committed:
new has method on States

martinheidegger committed:
different error for already finished requests

martinheidegger committed:
allowing only one request to be active

martinheidegger committed:
allowing members to cancel request

martinheidegger committed:
fixing bug where re-setting of a state would cause incosistencies

martinheidegger committed:
member can not cancel request by other member

martinheidegger committed:
denying a request

martinheidegger committed:
typo

martinheidegger committed:
refactor: extracting handleDeny and handleCancel

martinheidegger committed:
better error name

martinheidegger committed:
cant accept own request

martinheidegger committed:
adding test of accepting request

martinheidegger committed:
basic acceptance of requests

martinheidegger committed:
adding test to check that multiple members needed to add a request

martinheidegger committed:
making follow up requests active

martinheidegger committed:
can not remove a non-member

martinheidegger committed:
allowing to remove members

martinheidegger committed:
Preventing readding of members

martinheidegger committed:
Preventing to add already added members.

martinheidegger committed:
Making sure that the same request id is not used twice

martinheidegger committed:
Making sure that one signature less is required for remove requests

martinheidegger committed:
also returning that request

martinheidegger committed:
Allowing for all members to be removed

martinheidegger committed:
States: replacing object access with interface access.

martinheidegger committed:
cleaned state

martinheidegger committed:
started working on versioned state

RangerMauve committed:
Merge pull request #2 from consento-org/permissions

Extracted permission object

RangerMauve committed:
Progress on Sync and member refactor

martinheidegger committed:
removed swp file and fixing lint

martinheidegger committed:
updating dependencies and style fixes.

martinheidegger committed:
removed the eslint exceptions with easy fixes

RangerMauve committed:
Progress on refactor with new permissions code

RangerMauve committed:
More passing tests, more failing tests

RangerMauve committed:
Make tests pass

RangerMauve committed:
Merge remote-tracking branch 'origin' into initial

RangerMauve committed:
Update README

martinheidegger commented on src/Permissions.ts:
```diff
@@ -0,0 +1,269 @@
+  get currentMembers (): MemberId[] {
+    return [...this.members.values()]
+      .filter(({ state }) => state === 'added')
+      .sort(({ timestamp: time1 }, { timestamp: time2 }) => time1.compare(time2))
```
We could use a Timestamp.compare util here.

RangerMauve replied:
What advantage does that have over using `time1.compare(time2)`?

RangerMauve replied:
Either way it's removed now. :P

martinheidegger replied:
Oh, right: it doesn't have any advantage: My bad. sorry for confusion

martinheidegger commented on src/Permissions.ts:
```diff
@@ -0,0 +1,269 @@
+  readonly openRequests = new Map<RequestId, Request>()
+
+  get currentMembers (): MemberId[] {
+    return [...this.members.values()]
```
This maybe should be precomputed because this seems like a common operation?

RangerMauve replied:
Now gets pre-computed

martinheidegger commented on src/Permissions.ts:
```diff
@@ -0,0 +1,269 @@
+
+  private readonly memberTime = new Map<MemberId, Timestamp>()
+  private readonly openRequestsByMember = new Map<MemberId, Request[]>()
+  private readonly openRequestTime = new Map<RequestId, Timestamp>()
```
I read this as "When the request was opened" not as the "time of the open request".

RangerMauve replied:
Renamed, how does `latestRequestTimestamp` sound?

martinheidegger commented on src/Permissions.ts:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  private getMembersByState (state: MemberStateState): MemberId[] {
+    return [...this.members.values()]
```
Why is this not sorted by time?

RangerMauve replied:
Resolved by using versionedsate

martinheidegger commented on src/Permissions.ts:
```diff
@@ -0,0 +1,269 @@
+
+    const knownAtTime = this.getMembersByStateAndTime('added', requestTime)
+
+    // TODO: Use request timestamp to get members at that time
```
TODO is done ... can be removed.

martinheidegger commented on src/Permissions.ts:
```diff
@@ -0,0 +1,269 @@
+  private getMembersByStateAndTime (state: MemberStateState, timestamp: Timestamp): MemberId[] {
+    return [...this.members.values()]
+      .filter((member) => {
+        if (member.state !== state) return false
```
It seems like this logic doesn't consider a removed member that was added previously?

RangerMauve replied:
FN got replaced with versionedstate stuff

martinheidegger commented on src/Sync.ts:
```diff
@@ -0,0 +1,85 @@
+        hasProcessed = true
+        feed.increment()
+      } catch (e) {
+        if (String(e.message).startsWith('Response for unknown request')) {
```
Use error codes.

RangerMauve replied:
Can we merge this in before doing the error codes refactor?

martinheidegger committed:
updating dependencies

martinheidegger committed:
fix: lint

martinheidegger committed:
Fixing history approach to things (throwing out performance)

RangerMauve commented on src/VersionedStates.ts:
```diff
@@ -1,195 +1,140 @@
+      }
+      alreadyFound.add(entry.id)
+      if (entry.state !== undefined) {
+        entries.unshift([entry.id, entry.state, entry.timestamp])
```
Why does this need to be placed into an array instead of using `yield`?

martinheidegger replied:
Sorting.

RangerMauve replied:
Ah, is that so we can iterate from oldest to newest? `iterateWithTime` makes sense as a name.

RangerMauve committed:
Abstract member related code to MemberList class

martinheidegger commented on src/MemberList.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  private recalculate (): void {
+    this.lastAdded = this._state.byState('added')
```
As it is now: there is no precalculation happening. You should be able to just use byState and it will always be correct (trade: code brevity for performance)

RangerMauve committed:
Remove pre-calculate from MemberList

martinheidegger reviewed (approved):
We can merge it for now. more progress would be good to be made in issues.

RangerMauve committed:
Merge branch 'main' into initial
